### PR TITLE
[BE][Laravel] Create API to Signup users

### PR DESCRIPTION
**Pull request by Vali-R**
## Asana Link
https://app.asana.com/0/1202860323783357/1203223750808872/f

 ## Commands
```
php artisan migrate
php artisan db:seed
php artisan serve
```
**Please send POST request to `http://localhost:YOUR_PORT/api/signup`**

## Pre-conditions
-  **Please setup your db first**

-  **It expects body**

_example for body:_
```
{
    "first_name":"Test",
    "last_name":"User",
    "email":"test@gmail.com",
    "password":"test123"
}
```
-  **`Accept`  header needs to be set to `application/json`**

## Expected Output

- [x]  **Validation for the body values**
- [x] **Should be able to signup users**
- [x] **Should not be able to signup with the same email for the second time**  

## Documentation
- Postman APIs - https://grey-firefly-150180.postman.co/workspace/e-learning~b9d71430-452d-47cc-8c64-e2b2e960bd99/documentation/22825265-53a51ce8-b099-477e-9f5d-3a6884ffae89

## Notes
**Backend integration only**

 ## Screenshots

image

image

### CODE REVIEW

File: app/Http/Controllers/Controller.php [hunk 1]
```diff
@@ -9,5 +9,7 @@
 
 class Controller extends BaseController
 {
-    use AuthorizesRequests, DispatchesJobs, ValidatesRequests;
+    use AuthorizesRequests;
+    use DispatchesJobs;
+    use ValidatesRequests;
```
Comment (escbooster12-dev) [inline app/Http/Controllers/Controller.php:14]: - unnecessary changes. please reset this one.

File: app/Models/Follow.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+use Illuminate\Database\Eloquent\Model;
+
+class Follow extends Model
+{
+    use HasFactory;
+
+    public function user()
+    {
+        return $this->belongsTo(User::class);
+    }
```
Comment (escbooster12-dev) [inline app/Models/Follow.php:15]: - this would probably won't work. Please specify the `foreign_key`
- test the relationships if it really works (do to all related)

File: app/Models/Question.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+use Illuminate\Database\Eloquent\Model;
+
+class Question extends Model
+{
+    use HasFactory;
+
+    public function useranswer()
+    {
+        return $this->hasMany(UserAnswer::class);
+    }
```
Comment (escbooster12-dev) [inline app/Models/Question.php:15]: - use snake case for variable or function name (apply to all related)

File: app/Models/Quiz.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    public function userlesson()
+    {
+        return $this->hasMany(UserLesson::class);
+    }
+
+    public function question()
+    {
+        return $this->hasMany(Question::class);
+    }
```
Comment (escbooster12-dev) [inline app/Models/Quiz.php:20]: - use plural name for `many` relationship

File: app/Models/User.php [hunk 5]
```diff
@@ -41,4 +44,29 @@ class User extends Authenticatable
     protected $casts = [
         'email_verified_at' => 'datetime',
     ];
+
+    public function setPasswordAttribute($password)
+    {
+        $this->attributes['password']=bcrypt($password);
+    }
```
[the remark this answers is not in the document]
Comment (Vali-R) [reply app/Models/User.php:51]: It is actually from the laravel course, it seems good practice, I believe it is called `Eloquent mutator`, and makes sure the password is always hashed whenever its changed
Please check referenced the link to the course below (starting 2:44 min):
https://laracasts.com/series/laravel-8-from-scratch/episodes/46

[the remark this answers is not in the document]
Comment (escbooster12-dev) [reply app/Models/User.php:51]: @Vali-R may I ask some screenshots the data of your `users` table.

[the remark this answers is not in the document]
Comment (Vali-R) [reply app/Models/User.php:51]: @escbooster12-dev 
image

[the remark this answers is not in the document]
Comment (Vali-R) [reply app/Models/User.php:51]: first 5 are seeded users, that is why their hashed passwords are the same, but the other 2 are the one I create via API

File: routes/web.php [hunk 6]
```diff
@@ -15,4 +15,4 @@
 
 Route::get('/', function () {
     return view('welcome');
-});
+});
```
Comment (escbooster12-dev) [inline routes/web.php:18]: - missing newline

**Review by escbooster12-dev — APPROVED** [review]
